[#399] #IMPLEMENT Helper methods for database script execution

fabiogouw opened the pull request:
Methods to run scripts against databases:

- SQL scripts for MS SQL, MySQL, Oracle, Postgre;
- Lua scripts for Redis; 
- N1QL scripts for Couchbase

fabiogouw committed:
[#399] #IMPLEMENT 'assemblyName: DotNet.Testcontainers; function:
ExecScript
{Add function in the database modules to execute scripts}

HofmeisterAn commented:
I'll do the review in the next days. Did you check the test issue `ShouldExecuteCommandsWhenScriptExecuted`?

fabiogouw commented:
I'm still working on it, Andre, trying to figure it out.

fabiogouw commented:
> I'm still working on it, Andre, trying to figure it out.

I guess it's ok now!

HofmeisterAn commented:
> I guess it's ok now!

👍  I'll do the review end of next week. Unfortunately, I'm not able to do it earlier. Thanks.

HofmeisterAn commented on tests/DotNet.Testcontainers.Tests/Unit/Containers/Unix/Modules/Databases/CouchbaseTestcontainerTest.cs:
```diff
@@ -233,6 +233,25 @@ public void ShouldThrowArgumentOutOfRangeExceptionWhenClusterAnalyticsRamSizeIsN
     }
 
+    [Fact]
+    public async Task ShouldExecuteCommandsWhenScriptExecuted()
```
A consistent test method name (for all script tests) would be nice. What about something like `ExecScriptInRunningContainer`?

HofmeisterAn commented on tests/DotNet.Testcontainers.Tests/Unit/Containers/Unix/Modules/Databases/CouchbaseTestcontainerTest.cs:
```diff
@@ -233,6 +233,25 @@ public void ShouldThrowArgumentOutOfRangeExceptionWhenClusterAnalyticsRamSizeIsN
+      // Given
+      _ = await this.couchbaseFixture.Container.CreateBucket("MyBucket")
+        .ConfigureAwait(false);
+      var script = @"
```
Please use `const string script` instead of `var script`, other tests too.

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Modules/IScriptExecution.cs:
```diff
@@ -0,0 +1,19 @@
+  /// <summary>
+  /// Define the contract for database modules that has the script execution capability.
+  /// </summary>
+  public interface IScriptExecution
```
What about `IDatabaseScript`?

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Modules/IScriptExecution.cs:
```diff
@@ -0,0 +1,19 @@
+    /// <param name="scriptContent">The content of the script to be executed.</param>
+    /// <returns>Task that completes when the script has been executed.</returns>
+    [PublicAPI]
+    Task<ExecResult> ExecScript(string scriptContent);
```
`ExecScriptAsync`

HofmeisterAn commented on src/DotNet.Testcontainers/Containers/Modules/IScriptExecution.cs:
```diff
@@ -0,0 +1,19 @@
+  public interface IScriptExecution
+  {
+    /// <summary>
+    /// Executes a script in the database container.
```
`Executes a database script in the running Testcontainer.`

HofmeisterAn reviewed (changes requested):
Shouldn't we add the interface `IScriptExecution` to the base class `TestcontainerDatabase`? If the module does not support database scripts (or it has no implementation yet) we can simply throw `NotImplementedException`.

I suggest something like this:

```cs
// TestcontainerDatabase.cs

public abstract class TestcontainerDatabase : HostedServiceContainer, IDatabaseScript
{
  protected TestcontainerDatabase(ITestcontainersConfiguration configuration, ILogger logger)
    : base(configuration, logger)
  {
  }

  // ...

  public virtual string GetTempScriptFile()
  {
    return Path.Combine("/tmp", Guid.NewGuid().ToString("N"), ".tmp");
  }

  public virtual Task<ExecResult> ExecScriptAsync(string script)
  {
    throw new NotImplementedException();
  }
}
```

Then each supported module can override `ExecScriptAsync`, e. g.:

```cs
// PostgreSqlTestcontainer.cs

public override async Task<ExecResult> ExecScriptAsync(string script)
{
  var tempScriptFile = this.GetTempScriptFile();

  await this.CopyFileAsync(tempScriptFile, Encoding.UTF8.GetBytes(script))
    .ConfigureAwait(false);

  return await this.ExecAsync(new[] { "psql", "-U", this.Username, "-d", this.Database, "-a", "-f", tempScriptFile })
    .ConfigureAwait(false);
}
```

Everything else looks good 👍 . Thanks for the pull request and contribution.

fabiogouw commented:
> Shouldn't we add the interface `IScriptExecution` to the base class `TestcontainerDatabase`? If the module does not support database scripts (or it has no implementation yet) we can simply throw `NotImplementedException`.
> 
> I suggest something like this:
> 
> ```cs
> // TestcontainerDatabase.cs
> 
> public abstract class TestcontainerDatabase : HostedServiceContainer, IDatabaseScript
> {
>   protected TestcontainerDatabase(ITestcontainersConfiguration configuration, ILogger logger)
>     : base(configuration, logger)
>   {
>   }
> 
>   // ...
> 
>   public virtual string GetTempScriptFile()
>   {
>     return Path.Combine("/tmp", Guid.NewGuid().ToString("N"), ".tmp");
>   }
> 
>   public virtual Task<ExecResult> ExecScriptAsync(string script)
>   {
>     throw new NotImplementedException();
>   }
> }
> ```
> 
> Then each supported module can override `ExecScriptAsync`, e. g.:
> 
> ```cs
> // PostgreSqlTestcontainer.cs
> 
> public override async Task<ExecResult> ExecScriptAsync(string script)
> {
>   var tempScriptFile = this.GetTempScriptFile();
> 
>   await this.CopyFileAsync(tempScriptFile, Encoding.UTF8.GetBytes(script))
>     .ConfigureAwait(false);
> 
>   return await this.ExecAsync(new[] { "psql", "-U", this.Username, "-d", this.Database, "-a", "-f", tempScriptFile })
>     .ConfigureAwait(false);
> }
> ```
> 
> Everything else looks good 👍 . Thanks for the pull request and contribution.

@HofmeisterAn, thanks for your comments, I changed the code so most of them are implemented now, but there's one I'd like to discuss further.

When I was developing this code, my first thought was to change the abstract class TestcontainerDatabase so it could have the ExecScript method, like your suggestion. But I've found out that not all databases have the scripting capability (in the case, CouchDB). So it was strange to me having the CouchDbTestcontainer class with this method, even throwing the NotImplemented exception, so I decided to implement the IScriptExecution interface (now IDatabaseScript, I changed its name) on just the databases that have this feature (most of them, in fact). I think that if a class implements a given interface, it has to implement the feature.

One alternative to this case, for the databases that doesn't have a native script capability, is to let it run generic bash scripts. At least it would work for CouchDB, since it can be manipulated with http call (curl).

What do you think?

HofmeisterAn reviewed (commented):
> One alternative to this case, for the databases that doesn't have a native script capability, is to let it run generic bash scripts. At least it would work for CouchDB, since it can be manipulated with http call (curl).

I thought of that, too. If a database does not have native script capabilities `sh` would be helpful. The modules should be moved into their own dependencies anyway. Then they can provide database specific features much easier. As long as this is not the case I think it's easier to maintain from the base class.

> even throwing the NotImplemented exception

I think it's not that bad. If someone needs the script capability then they can always create a pull request. Modules are very simple right now.

fabiogouw commented:
Sounds ok to me! I'll implement the `sh` capability in the base class and move the reference to the IDatabaseScript interface to the base class.

HofmeisterAn reviewed (approved):
👍 Thanks, I'll merge it ~~tomorrow~~.

**Edit**:
Sorry - I can't do it today. Tomorrow, I promise 🤞.

HofmeisterAn commented:
`1.6.0-beta.2028`

fabiogouw commented:
Hello, @HofmeisterAn ! I've noticed that you released a new version of the lib (1.6.0). Is this PR included?

HofmeisterAn commented:
> Hello, @HofmeisterAn ! I've noticed that you released a new version of the lib (1.6.0). Is this PR included?

Yes, it is. I'm sorry — I forgot to add it to the release notes. The issue wasn't assigned to the milestone. I'll update the information later that day.

https://github.com/testcontainers/testcontainers-dotnet/blob/2aeec1e9b88b8580c9914b73bee56583ad005e03/src/DotNet.Testcontainers/Containers/Modules/IDatabaseScript.cs#L19

https://github.com/testcontainers/testcontainers-dotnet/blob/2aeec1e9b88b8580c9914b73bee56583ad005e03/src/DotNet.Testcontainers/Containers/Modules/Databases/MsSqlTestcontainer.cs#L34

fabiogouw commented:
Great, thanks!
There's also this PR #392, if you could mention it too, the release notes description will be more complete.
